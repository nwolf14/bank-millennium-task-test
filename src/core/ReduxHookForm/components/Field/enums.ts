export enum InputValueType {
  boolean,
  number,
  string,
}
