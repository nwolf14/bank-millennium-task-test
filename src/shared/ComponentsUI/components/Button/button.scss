@import './styles/button-variants.scss';
@import '../../../../styles/mixins/button/button.scss';
@import '../../../../styles/mixins/button/button-sizes.scss';

.Button {
  @include button-variants;
  @include button-size;
  @include button;
  border-radius: 4px;

  &--full-width {
    width: 100%;
  }

  &__icon {
    &--small {
      height: 12px;
    }

    &--medium {
      height: 14px;
    }

    &--large {
      height: 16px;
    }

    &--start {
      margin-right: 10px;
    }

    &--end {
      margin-left: 10px;
    }
  }
}
