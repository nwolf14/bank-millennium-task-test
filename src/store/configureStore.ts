import createSagaMiddleware from 'redux-saga';
import { applyMiddleware, createStore, EmptyObject, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { routerMiddleware } from 'connected-react-router';

// others
import { TAnyAction, TMainState } from '../types';

// store
import reducers from './reducers';
import sagaMiddlewareRuns from './watchers';

// utils
import { history } from '../utils/history';

const sagaMiddleware = createSagaMiddleware();

const configureStore = (): Store<EmptyObject & TMainState, TAnyAction<any>> & {
  dispatch: unknown;
} => {
  const store = createStore(
    reducers(history),
    {},
    composeWithDevTools(
      applyMiddleware(routerMiddleware(history), sagaMiddleware)
    )
  );
  sagaMiddlewareRuns(sagaMiddleware);

  return store;
};

export default configureStore;
