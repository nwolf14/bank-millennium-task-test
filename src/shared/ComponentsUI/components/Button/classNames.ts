import mapValues from 'lodash/mapValues';

// others
import { Color, Size, Variant } from './enums';

export const className = 'Button';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      fullwidth: `${className}--full-width`,
      forcedHover: {
        ...mapValues(Variant, (variant) => ({
          ...mapValues(
            Color,
            (color) => `${className}--${variant}--${color}--forced-hover`
          ),
        })),
      },
      ...{
        ...mapValues(Color, (color) => ({
          ...mapValues(
            Variant,
            (variant) => `${className}--${variant}--${color}`
          ),
        })),
      },
      ...{ ...mapValues(Size, (size) => `${className}--${size}`) },
      ...{ ...mapValues(Variant, (variant) => `${className}--${variant}`) },
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      start: `${className}__icon--start`,
      end: `${className}__icon--end`,
      ...{ ...mapValues(Size, (size) => `${className}__icon--${size}`) },
    },
  },
};
