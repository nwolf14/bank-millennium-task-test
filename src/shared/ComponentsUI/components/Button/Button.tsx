import classnames from 'classnames';
import { FC, MouseEvent, ReactNode } from 'react';

// hooks
import { useRippleEffect } from '../../../../hooks/useRippleEffect/useRippleEffect';

// others
import { className as classNameButton, classNames } from './classNames';
import { Color, Size, Variant } from './enums';

// styles
import './button.scss';

export type TProps = {
  children?: ReactNode | string;
  className?: string;
  color?: Color;
  disabled?: boolean;
  disableRippleEffect?: boolean;
  endIcon?: string;
  forcedHover?: boolean;
  fullWidth?: boolean;
  onClick?: () => void;
  size?: Size;
  startIcon?: string;
  style?: { [key: string]: number | string };
  type?: 'button' | 'reset' | 'submit';
  variant?: Variant;
};

export const Button: FC<TProps> = ({
  children,
  className = '',
  color = Color.primary,
  disabled = false,
  disableRippleEffect = false,
  endIcon,
  forcedHover = false,
  fullWidth = false,
  onClick,
  size = Size.medium,
  startIcon,
  style = {},
  type = 'button',
  variant = Variant.contained,
}) => {
  const { rippleEffect, triggerRippleEffect } = useRippleEffect(
    classNames[classNameButton].modificators[color][variant]
  );

  const Icon = ({
    placement,
    src,
  }: {
    placement: 'start' | 'end';
    src: string;
  }): JSX.Element => (
    <img
      alt="icon"
      className={classnames(
        classNames.icon.name,
        classNames.icon.modificators[placement],
        classNames.icon.modificators[size]
      )}
      src={src}
    />
  );

  const onClickHandler = (event: MouseEvent<HTMLButtonElement>): void => {
    if (!disableRippleEffect) {
      triggerRippleEffect(event);
    }

    if (onClick) {
      onClick();
    }
  };

  return (
    <button
      className={classnames(
        classNames[classNameButton].name,
        { [classNames[classNameButton].modificators.fullwidth]: fullWidth },
        {
          [classNames[classNameButton].modificators.forcedHover[variant][
            color
          ]]: forcedHover,
        },
        classNames[classNameButton].modificators[size],
        classNames[classNameButton].modificators[variant],
        classNames[classNameButton].modificators[color][variant],
        className
      )}
      disabled={disabled}
      onClick={onClickHandler}
      style={style}
      type={type}
    >
      {startIcon && <Icon placement="start" src={startIcon} />}
      {children}
      {endIcon && <Icon placement="end" src={endIcon} />}
      {rippleEffect}
    </button>
  );
};

export default Button;
