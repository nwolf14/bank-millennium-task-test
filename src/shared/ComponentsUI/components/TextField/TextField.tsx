import camelCase from 'lodash/camelCase';
import classnames from 'classnames';
import { ChangeEvent, forwardRef, useEffect, useState } from 'react';

// components
import ValidationMessage from '../../../ValidationMessage/ValidationMessage';
import TextFieldInputWrapper, {
  TProps as TTextFieldInputWrapperProps,
} from './components/TextFieldInputWrapper/TextFieldInputWrapper';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameTextField, classNames } from './classNames';
import { HEIGHTS } from './constants';
import { Margin } from '../../enums';
import { Size } from './enums';
import { INPUT_STATES_INITIAL } from '../../constants';
import { THookType as TMetaProps } from '../../../../core/ReduxHookForm/components/Field/hooks/useMetaProps';
import { TObject } from '../../../../types';

// styles
import './text-field.scss';

// utils
import { getErrorMessage } from '../../../../utils/form/inputStates';

export type TProps = Omit<
  TTextFieldInputWrapperProps,
  | 'error'
  | 'height'
  | 'inputStates'
  | 'multiline'
  | 'onBlur'
  | 'onChange'
  | 'onFocus'
  | 'required'
  | 'rows'
  | 'setInputStates'
  | 'size'
  | 'value'
> & {
  className?: string;
  errorMessage?: string;
  defaultValue?: boolean | number | string | Array<string>;
  fullWidth?: boolean;
  helperText?: string;
  margin?: Margin;
  maxRows?: number;
  meta?: TMetaProps;
  minRows?: number;
  multiline?: boolean;
  onBlur?: (value: number | string) => void;
  onChange?: (value: number | string) => void;
  onFocus?: (value: number | string) => void;
  required?: boolean;
  size?: Size;
  style?: TObject<number | string>;
  value?: number | string;
};

export const TextField = forwardRef<HTMLInputElement, TProps>(
  (
    {
      className = '',
      defaultValue = '',
      errorMessage = '',
      fullWidth = false,
      helperText,
      margin = Margin.marginBottom,
      meta = {},
      multiline = false,
      required = false,
      maxRows = 0,
      minRows = 0,
      onBlur = null,
      onChange = null,
      onFocus = null,
      size = Size.medium,
      style = {},
      value: controlledValue = null,
      ...restProps
    },
    ref
  ) => {
    const [height, setHeight] = useState(`${HEIGHTS[size] * (minRows || 1)}px`);
    const [inputStates, setInputStates] = useState(INPUT_STATES_INITIAL);
    const [rows, setRows] = useState(minRows || 1);
    const [value, setValue] = useState(defaultValue);
    const { classNamesWithTheme } = useTheme(classNames);
    const error = getErrorMessage(meta as TMetaProps) || errorMessage;

    const setInputHeight = (value: string): void => {
      const amountLines = value.split(/\r?\n|\r/).length;

      if (!minRows && !maxRows) {
        setRows(amountLines);
        setHeight(`${HEIGHTS[size] * amountLines}px`);
      } else if (amountLines <= maxRows) {
        if (amountLines <= minRows) {
          setHeight(`${HEIGHTS[size] * minRows}px`);
        } else {
          setHeight(`${HEIGHTS[size] * amountLines}px`);
        }
      }
    };

    const onBlurHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setInputStates({ ...inputStates, focus: false });

      if (onBlur) {
        onBlur(value);
      }
    };

    const onChangeHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setValue(value);

      if (multiline) {
        setInputHeight(value);
      }

      if (onChange) {
        onChange(value);
      }
    };

    const onFocusHandler = (event: ChangeEvent<HTMLInputElement>): void => {
      const { value } = event.target;

      setInputStates({ ...inputStates, focus: true });

      if (onFocus) {
        onFocus(value);
      }
    };

    const getValue = (): string | number | boolean | Array<string> =>
      controlledValue !== null ? controlledValue : value;

    useEffect(() => {
      if (getValue() && multiline) {
        setInputHeight(defaultValue as string);
      }
    }, []);

    return (
      <div
        className={classnames(
          classNamesWithTheme[classNameTextField].name,
          {
            [classNamesWithTheme[classNameTextField].modificators.fullWidth]:
              fullWidth,
          },
          classNamesWithTheme[classNameTextField].modificators[
            camelCase(margin)
          ],
          className
        )}
        style={style}
      >
        <TextFieldInputWrapper
          error={error}
          height={height}
          inputStates={inputStates}
          multiline={multiline}
          onBlur={onBlurHandler}
          onChange={onChangeHandler}
          onFocus={onFocusHandler}
          ref={ref}
          required={required}
          rows={rows}
          setInputStates={setInputStates}
          size={size}
          value={getValue() as string | number}
          {...restProps}
        />
        <ValidationMessage error={error} helperText={helperText} />
      </div>
    );
  }
);

export default TextField;
