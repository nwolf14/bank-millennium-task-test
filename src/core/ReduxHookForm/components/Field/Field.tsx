import { defer } from 'lodash';
import {
  FC,
  forwardRef,
  ForwardRefExoticComponent,
  ReactNode,
  RefObject,
  useEffect,
} from 'react';
import { useDispatch, useSelector } from 'react-redux';

// hooks
import { useInputProps } from './hooks/useInputProps';
import {
  THookType as TFieldMetaProps,
  useMetaProps,
} from './hooks/useMetaProps';
import { useValidators } from './hooks/useValidators';

// others
import { TAsyncValidator, TSyncValidator } from '../../types';
import { TProps as TTextFieldProps } from '../../../../shared/ComponentsUI/components/TextField/TextField';
import { TProps as TCheckboxProps } from '../../../../shared/ComponentsUI/components/Checkbox/Checkbox';
import { TProps as TRadioButtonProps } from '../../../../shared/ComponentsUI/components/RadioButton/RadioButton';
import { TProps as TSwitchProps } from '../../../../shared/ComponentsUI/components/Switch/Switch';

// store
import { getFieldSelectorCreator } from '../../../../store/reduxHookForm/selectors';
import { initField } from '../../../../store/reduxHookForm/actions';
import { InputValueType } from './enums';

export type TProps = Partial<TCheckboxProps> &
  Partial<TRadioButtonProps> &
  Partial<TSwitchProps> &
  Partial<TTextFieldProps> & {
    afterSubmit?: () => void;
    asyncValidators?: Array<TAsyncValidator>;
    beforeSubmit?: () => void;
    children?: (
      metaProps: Partial<TFieldMetaProps>,
      restProps?: any
    ) => ReactNode;
    Component?: FC<any> | ForwardRefExoticComponent<any>;
    data?: { [key: string]: any };
    formatOnBlur?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnChange?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnFocus?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    formatOnInit?: (
      value: boolean | number | string | Array<string>
    ) => boolean | number | string;
    formName?: string;
    inputValueType?: InputValueType;
    name: string;
    parse?: (
      value: boolean | number | string | Array<string>,
      name: string
    ) => boolean | number | string;
    ref?: RefObject<HTMLInputElement>;
    subscriptionFields?: Array<string>;
    syncValidators?: Array<TSyncValidator>;
    touched?: boolean;
    touchedAfterClick?: boolean;
    visited?: boolean;
  };

const Field = forwardRef<HTMLInputElement, TProps>(
  (
    {
      afterSubmit,
      asyncValidators = [],
      beforeSubmit,
      children,
      Component,
      data,
      defaultValue,
      formatOnBlur,
      formatOnChange,
      formatOnFocus,
      formatOnInit,
      formName = '',
      inputValueType = InputValueType.string,
      name = '',
      parse,
      subscriptionFields = [],
      syncValidators = [],
      touched: initialTouched = false,
      touchedAfterClick = false,
      visited: initialVisited = false,
      ...restProps
    },
    ref
  ) => {
    const field = useSelector(getFieldSelectorCreator(formName, name));
    const { touched, value = '', visited } = field || {};
    const dispatch = useDispatch();
    const inputProps = useInputProps(formName, name);
    const metaProps = useMetaProps(formName, name);
    const { getSyncErrors, updateAsyncValidators, updateSyncValidators } =
      useValidators(
        asyncValidators,
        formName,
        name,
        subscriptionFields,
        syncValidators
      );
    const wrapperRestProps = {
      ...inputProps,
      ...restProps,
      ...(ref ? { ref: ref } : {}),
    };

    const getInitialValue = (): boolean | number | string | Array<string> => {
      const initValue = formatOnInit
        ? formatOnInit(defaultValue || '')
        : defaultValue;

      switch (inputValueType) {
        case InputValueType.boolean:
          return initValue || false;
        case InputValueType.number:
          return initValue || 0;
        default:
          return initValue || '';
      }
    };

    useEffect(() => {
      if (formName) {
        const initialValue = getInitialValue();

        defer(() => {
          dispatch(
            initField({
              field: {
                afterSubmit,
                active: false,
                asyncErrors: [],
                beforeSubmit,
                data,
                formatOnBlur,
                formatOnChange,
                formatOnFocus,
                initialValue: value,
                isPending: false,
                parse,
                syncErrors: getSyncErrors(initialValue),
                touched: initialTouched,
                touchedAfterClick,
                value: initialValue,
                visited: initialVisited,
              },
              formName,
              name,
            })
          );
        });
      }
    }, [formName]);

    useEffect(() => {
      if (field) {
        updateAsyncValidators(value);
        updateSyncValidators(value);
      }
    }, [touched, value, visited]);

    if (children) {
      return <>{children(metaProps, wrapperRestProps)}</>;
    }

    if (Component) {
      return <Component meta={metaProps} {...wrapperRestProps} />;
    }

    return null;
  }
);

export default Field;
