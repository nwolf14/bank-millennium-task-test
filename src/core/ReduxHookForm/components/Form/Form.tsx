import forOwn from 'lodash/forOwn';
import {
  FunctionComponent,
  memo,
  ReactElement,
  ReactNode,
  useEffect,
  useMemo,
} from 'react';
import { useDispatch, useSelector } from 'react-redux';

// components
import Field from '../Field/Field';

// others
import { TField } from '../../../../store/reduxHookForm/types';
import { TObject } from '../../../../types';

// store
import {
  destroyForm,
  mountForm,
  submit,
  updateForm,
} from '../../../../store/reduxHookForm/actions';
import { getFieldsSelectorCreator } from '../../../../store/reduxHookForm/selectors';
import { UPDATE_FORM_VALIDATOR } from '../../../../store/reduxHookForm/actionsType';

// utils
import { getChildrenWithForwardedProps } from '../../../../utils/react-children/getChildrenWithForwardedProps';
import { markInputsAsTouched } from '../../utils/markInputsAsTouched';

export type TProps = {
  asyncTimeDelay?: number;
  children: ReactNode;
  formName: string;
  isValid?: boolean;
  onSubmit: (formData: {
    [key: string]: boolean | number | string | Array<string>;
  }) => void;
  validate?: (fields: { [key: string]: TField }) => boolean;
};
const Form: FunctionComponent<TProps> = ({
  asyncTimeDelay = 0,
  children,
  formName,
  isValid: isValidInitial = false,
  onSubmit,
  validate = () => true,
}) => {
  const dispatch = useDispatch();
  const content = useMemo(
    () =>
      getChildrenWithForwardedProps(
        children as ReactElement,
        { formName },
        Field
      ),
    [children, formName]
  );
  const fields = useSelector(getFieldsSelectorCreator(formName));

  const onValidateHandler = (): void => {
    const isValid = validate(fields as TObject<TField>);

    dispatch(
      updateForm({ form: { isValid }, formName }, UPDATE_FORM_VALIDATOR)
    );
  };

  const onSubmitHandler = (event: any): void => {
    event.preventDefault();

    dispatch(submit(formName));
    onSubmit(getFieldsValues());
    markInputsAsTouched(dispatch, formName, fields as TObject<TField>);
    onValidateHandler();
  };

  const getFieldsValues = () => {
    const formData: {
      [key: string]: boolean | number | string | Array<string>;
    } = {};

    forOwn(fields, (field: TField, name: string) => {
      const { parse, value } = field;
      formData[name] = parse ? parse(value, name) : value;
    });

    return formData;
  };

  useEffect(() => {
    dispatch(
      mountForm({
        [formName]: {
          asyncTimeDelay,
          error: '',
          fields: {},
          isPending: false,
          isValid: isValidInitial,
        },
      })
    );

    return () => {
      dispatch(destroyForm(formName));
    };
  }, []);

  return <form onSubmit={onSubmitHandler}>{content}</form>;
};

const memoizedForm = memo(Form, (prevProps: TProps, props: TProps): boolean => {
  /* istanbul ignore next */
  if (prevProps !== props) {
    /* istanbul ignore next */
    return true;
  }
  /* istanbul ignore next */
  return false;
});

export default memoizedForm;
