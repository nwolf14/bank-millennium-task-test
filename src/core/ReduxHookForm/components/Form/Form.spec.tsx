import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';

// components
import Form from './Form';
import Field from '../Field/Field';
import TextField from '../../../../shared/ComponentsUI/components/TextField/TextField';
import Button from '../../../../shared/ComponentsUI/components/Button/Button';

const stateMock = {
  reduxHookForm: {
    testForm: {
      asyncTimeDelay: 0,
      error: '',
      fields: {
        testField: {
          active: false,
          asyncErrors: [],
          initialValue: '',
          isPending: false,
          parse: (value: any) => value,
          syncErrors: [],
          touched: false,
          touchedAfterClick: false,
          value: '',
          visited: false,
        },
      },
      isPending: false,
      isValid: false,
    },
  },
};

describe('Form', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment, getByText } = render(
      <Provider store={mockStore(stateMock)}>
        <Form onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" />
          <Button type="submit">Click</Button>
        </Form>
      </Provider>
    );
    fireEvent.click(getByText('Click'));

    expect(asFragment()).toMatchSnapshot();
  });
});
