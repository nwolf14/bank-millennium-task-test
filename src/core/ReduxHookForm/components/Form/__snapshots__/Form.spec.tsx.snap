// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="TextField TextField--margin-bottom"
    >
      <div
        class="TextFieldInputWrapper TextFieldInputWrapper--medium"
      >
        <input
          class="TextFieldInput TextFieldInput--medium"
          data-testid="input"
          name="testField"
          placeholder=""
          style="height: 24px;"
          type="text"
          value=""
        />
        <label
          class="TextFieldLabel TextFieldLabel--medium"
        />
        <div
          class="TextFieldFilling"
        >
          <legend
            class="TextFieldFilling__legend"
          >
             
          </legend>
        </div>
      </div>
      <div
        class="ValidationMessage"
      />
    </div>
    <button
      class="Button Button--medium Button--contained Button--contained--primary"
      type="submit"
    >
      Click
      <span
        class="Button--contained--primary--ripple"
        style="left: 0px; top: 0px;"
      />
    </button>
  </form>
</DocumentFragment>
`;
